esp32/mods/modhttp.c

```diff
@@ -0,0 +1,1475 @@
+/*
+ * Copyright (c) 2019, Pycom Limited.
+ *
+ * This software is licensed under the GNU GPL version 3 or any
+ * later version, with permitted additional terms. For more information
+ * see the Pycom Licence v1.0 document supplied with this file, or
+ * available at https://www.pycom.io/opensource/licensing
+ */
+
+#include "py/mpconfig.h"
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/gc.h"
+
+#include "mpthreadport.h"
+#include "mpirq.h"
+
+#include "esp_err.h"
+#include "esp_https_server.h"
+#include "esp_http_client.h"
+#include "modhttp.h"
+#include "pycom_general_util.h"
+
+/******************************************************************************
+ DEFINE CONSTANTS
+ ******************************************************************************/
+
+#define MOD_HTTP_GET        (1)
+#define MOD_HTTP_PUT        (2)
+#define MOD_HTTP_POST       (4)
+#define MOD_HTTP_DELETE     (8)
+
+#define MOD_HTTP_MEDIA_TYPE_TEXT_XML        "text/xml"
+#define MOD_HTTP_MEDIA_TYPE_TEXT_PLAIN      "text/plain"
+#define MOD_HTTP_MEDIA_TYPE_APP_XML         "application/xml"
+
+
+#define MOD_HTTP_MEDIA_TYPE_TEXT_HTML_ID    (0)
+#define MOD_HTTP_MEDIA_TYPE_TEXT_XML_ID     (1)
+#define MOD_HTTP_MEDIA_TYPE_TEXT_PLAIN_ID   (2)
+#define MOD_HTTP_MEDIA_TYPE_APP_JSON_ID     (3)
+#define MOD_HTTP_MEDIA_TYPE_APP_OCTET_ID    (4)
+#define MOD_HTTP_MEDIA_TYPE_APP_XML_ID      (5)
+
+#define MOD_DEFAULT_HTTP_USER_AGENT         "ESP32 HTTP Client/1.0"
+#define MOD_HTTP_MAX_PAYLOAD_LEN         	(32768)
+
+
+/******************************************************************************
+ DEFINE PRIVATE TYPES
+ ******************************************************************************/
+typedef struct mod_http_resource_obj_s {
+	mp_obj_base_t base;
+	struct mod_http_resource_obj_s* next;
+	const char* uri;
+	uint8_t* value;
+	uint16_t value_len;
+	uint8_t mediatype;
+}mod_http_resource_obj_t;
+
+typedef struct mod_http_server_obj_s {
+	httpd_handle_t server;
+	mod_http_resource_obj_t* resources;
+	int max_resources;
+	int num_of_resources;
+}mod_http_server_obj_t;
+
+// From esp_httpd_priv.h
+#define HTTPD_SCRATCH_BUF  MAX(HTTPD_MAX_REQ_HDR_LEN, HTTPD_MAX_URI_LEN)
+
+struct httpd_req_aux {
+    struct sock_db *sd;                             /*!< Pointer to socket database */
+    char            scratch[HTTPD_SCRATCH_BUF + 1]; /*!< Temporary buffer for our operations (1 byte extra for null termination) */
```

Comment:
it does not make any differences but please prefer using types like uint8_t or uint16_t instead of the generic C types.